# Basics to Generalized Additive Models


#### Resources 
* [Generalized Additive Models in Python](https://pygam.readthedocs.io)
* [GAM: The Predictive Modeling Silver Bullet - Kim Larsen](https://multithreaded.stitchfix.com/blog/2015/07/30/gam/)
* [Regularization](https://multithreaded.stitchfix.com/assets/files/gam.pdf)
* [Medium - Getting Started with Generalized Additive Models in Python](https://codeburst.io/pygam-getting-started-with-generalized-additive-models-in-python-457df5b4705f)


### Import Libraries and Load Breast Cancer Dataset

In [4]:
import pandas as pd
from pygam import LogisticGAM
from sklearn.datasets import load_breast_cancer

data = load_breast_cancer()
df = pd.DataFrame(data.data, columns=data.feature_names)[['mean radius', 'mean texture', 'mean perimeter', 'mean area','mean smoothness', 'mean compactness']]
target_df = pd.Series(data.target)
df.describe()

mean radius  mean texture  mean perimeter    mean area  \
count   569.000000    569.000000      569.000000   569.000000   
mean     14.127292     19.289649       91.969033   654.889104   
std       3.524049      4.301036       24.298981   351.914129   
min       6.981000      9.710000       43.790000   143.500000   
25%      11.700000     16.170000       75.170000   420.300000   
50%      13.370000     18.840000       86.240000   551.100000   
75%      15.780000     21.800000      104.100000   782.700000   
max      28.110000     39.280000      188.500000  2501.000000   

       mean smoothness  mean compactness  
count       569.000000        569.000000  
mean          0.096360          0.104341  
std           0.014064          0.052813  
min           0.052630          0.019380  
25%           0.086370          0.064920  
50%           0.095870          0.092630  
75%           0.105300          0.130400  
max           0.163400          0.345400

With the breast cancer dataset we will have:
- 569 obserfations 
- 30 features


## Logistic GAM
Lets see what we can do with building out the model

Logistic GAM - https://pygam.readthedocs.io/en/latest/api/logisticgam.html

In [7]:
X = df[['mean radius', 'mean texture', 'mean perimeter', 'mean area',
        'mean smoothness', 'mean compactness']]
y = target_df

# fit model
gam = LogisticGAM().fit(X, y)
gam

LogisticGAM(callbacks=[Deviance(), Diffs(), Accuracy()], 
   fit_intercept=True, max_iter=100, 
   terms=s(0) + s(1) + s(2) + s(3) + s(4) + s(5) + intercept, 
   tol=0.0001, verbose=False)

Lets take a look at some of the summary statistics from the GAM classification model

In [8]:
gam.summary()

LogisticGAM                                                                                               
=============================================== ==========================================================
Distribution:                      BinomialDist Effective DoF:                                     19.4476
Link Function:                        LogitLink Log Likelihood:                                   -54.0256
Number of Samples:                          569 AIC:                                              146.9464
                                                AICc:                                             148.5483
                                                UBRE:                                               2.2856
                                                Scale:                                                 1.0
                                                Pseudo R-Squared:                                   0.8562
Feature Function                  Lam

/home/redne/.local/lib/python3.7/site-packages/ipykernel_launcher.py:1: UserWarning: KNOWN BUG: p-values computed in this summary are likely much smaller than they should be. 
 
Please do not make inferences based on these values! 

Collaborate on a solution, and stay up to date at: 
github.com/dswah/pyGAM/issues/163 

  """Entry point for launching an IPython kernel.


In [9]:
gam.accuracy(X, y)

0.9560632688927944